accounts/abi/type.go

```diff
@@ -121,18 +130,44 @@ func NewType(t string, internalType string, components []ArgumentMarshaling) (ty
 	// varSize is the size of the variable
 	var varSize int
 	if len(parsedType[3]) > 0 {
+		if !map[string]bool{"int": true, "uint": true, "fixed": true, "ufixed": true, "bytes": true}[parsedType[1]] {
+			return Type{}, fmt.Errorf("only int, uint, fixed, ufixed, bytes types can have variable size specified: %s", t)
+		}
 		var err error
-		varSize, err = strconv.Atoi(parsedType[2])
+		varSize, err = strconv.Atoi(parsedType[3])
 		if err != nil {
 			return Type{}, fmt.Errorf("abi: error parsing variable size: %v", err)
 		}
+		if strconv.Itoa(varSize) != parsedType[3] || varSize == 0 ||
+			(map[string]bool{"int": true, "uint": true, "fixed": true, "ufixed": true}[parsedType[1]] &&
+				(varSize == 0 || varSize%8 != 0 || varSize > 256)) ||
+			(parsedType[1] == "bytes" && varSize > 32) {
+			return Type{}, fmt.Errorf("abi: wrong variable size: %s", parsedType[3])
+		}
```

Comment:
This chunk is _really_ difficult to review -- and also, due to being complicated, it checks some things several times (like varSize ==0). Instead of inlining lookup maps, I think it would be clearer to turn the whole thing into a switch, e.g. 
```
	switch {
	case strconv.Itoa(varSize) != parsedType[3]:
		fallthrough
	case varSize == 0:
		fallthrough
	case varSize > 256:
		fallthrough
	case parsedType[1] == "int" && varSize%8 != 0:
		fallthrough
	case parsedType[1] == "uint" && varSize%8 != 0:
		fallthrough
	case parsedType[1] == "fixed" && varSize%8 != 0:
		fallthrough
	case parsedType[1] == "ufixed" && varSize%8 != 0:
		fallthrough
	case parsedType[1] == "bytes" && varSize > 32:
		return Type{}, fmt.Errorf("abi: wrong variable size: %s", parsedType[3])
	}
```

Reply:
Fixed. (I kept the lookup map though as it should be much more efficient than string comparisons)